### PR TITLE
add benchmark code

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
- add the following benchmark code
  - `ApiCallProtocolBenchmark`: api calls using different protocols (xml, json, ec2, query)
  - `ApiCallHttpClientBenchmark`: api calls using different http clients (urlconnection, apache)
  - `ClientCreationBenchmark`: create clients for cold start
- add benchmark buildspec file.
  - `ApiCallProtocolBenchmark` and `ClientCreationBenchmark` will be triggered whenever there's a change on `core` module
  - `ApiCallHttpClientBenchmark` will  be triggered whenever there's a change on `http-clients` module

**NOTES**:
- The total time to run all three benchmarks is around 20 minutes.
- Currently, the baseline scores are passed by the environment variables set on CodeBuild and it is not easy to maintain and lacks flexibility. I'm considering to use a json file to store the baseline result instead. This is open for discussion. 

**TODOs**:
- Clean up the commented-out lines
- Remove unnecessary code changes.

## Testing
running benchmark on codebuild

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallProtocolBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+@State(Scope.Thread)
+@Warmup(iterations = 3, time = 15, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 10, timeUnit = TimeUnit.SECONDS)
+@Fork(2) // To reduce difference between each run
+@BenchmarkMode(Mode.Throughput)
+public class ApiCallProtocolBenchmark {
+
+    private WireMockServer wireMockServer = new WireMockServer(options().port(PORT_NUMBER));
+
+    @Param( {"xml", "json", "ec2", "query"})
```
Comment (dagnir) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallProtocolBenchmark.java]: nit: put these in an `enum`

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java [hunk 2]
```diff
@@ -86,7 +86,7 @@
  * powerful synchronous client that adds an extra dependency and additional startup latency in exchange for more functionality,
  * like support for HTTP proxies.
  *
- * <p>See software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient for an alternative implementation.</p>
+ * <p>See software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient for an alternative implementation.</p> test
```
Comment (dagnir) [inline http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: nit: not sure what the " test" is

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentials.java [hunk 3]
```diff
@@ -31,7 +31,7 @@
 public interface AwsCredentials {
 
     /**
-     * Retrieve the AWS access key, used to identify the user interacting with AWS.
+     * Retrieve the AWS access key, used to identify the user interacting with AWS.test
```
Comment (dagnir) [inline core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentials.java]: Was put here by mistake?

File: buildspecs/runBenchmarks.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
+cd test/sdk-benchmarks
+BENCH_MARKS=()
+if git diff --name-only master HEAD | grep -q "http-clients/"; then
+    echo "Found changes in http-clients module, will run ApiCallHttpClientBenchmark and ClientCreationBenchmark"
+    BENCH_MARKS+=(ApiCallHttpClientBenchmark.run.UrlConnectionHttpClient:$URL_CONNECTION_CLIENT_BASELINE_SCORE)
+    BENCH_MARKS+=(ApiCallHttpClientBenchmark.run.ApacheHttpClient:$APACHE_CLIENT_BASELINE_SCORE)
+fi
+
+if git diff --name-only master HEAD | grep -q "core/"; then
```
Comment (dagnir) [inline buildspecs/runBenchmarks.sh]: This can be shortened with `git diff master core`

File: buildspecs/runBenchmarks.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
```
Comment (dagnir) [inline buildspecs/runBenchmarks.sh]: minor: FYI the CodeBuild Docker images all include Python if Bash wasn't the first choice

File: buildspecs/runBenchmarks.sh (hunk 4 above)
Comment (dagnir) [reply buildspecs/runBenchmarks.sh]: Believe we should also be using `--exit-code` here:

> Make the program exit with codes similar to diff(1). That is, it exits with 1 if there were differences and 0 means no differences.

File: http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java (hunk 2 above)
Comment (zoewangg) [reply http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/ApacheHttpClient.java]: This was added to trigger the CodeBuild during my test. Will remove when the PR is close to be merged.

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentials.java (hunk 3 above)
Comment (zoewangg) [reply core/auth/src/main/java/software/amazon/awssdk/auth/credentials/AwsCredentials.java]: Same as above, this was added to trigger the protocol tests benchmark. Will remove when the PR is close to be merged.

File: buildspecs/runBenchmarks.sh (hunk 4 above)
Comment (zoewangg) [reply buildspecs/runBenchmarks.sh]: +1

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    @TearDown(Level.Trial)
+    public void tearDown() {
+        wireMockServer.stop();
+    }
+
+    @Benchmark
+    public void run() {
+        runnable.run();
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java]: Sure, will do

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/coldstart/ClientCreationBenchmark.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Benchmark for creating the clients
+ */
+@BenchmarkMode(Mode.SingleShotTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 5)
+@Measurement(iterations = 5)
+@Fork(3)
+public class ClientCreationBenchmark {
```
[the remark this answers is not in the document]
Comment (dagnir) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/coldstart/ClientCreationBenchmark.java]: Also does this properly account for statics that only loaded once per process? i.e. any thing like `static final ObjectMapper`?

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallProtocolBenchmark.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    private Runnable runnable;
+
+    @Setup(Level.Trial)
+    public void setup() {
+        wireMockServer.start();
+
+        configureFor("localhost", PORT_NUMBER);
+        URI uri = getUri();
+
+        switch (protocol) {
```
Comment (dagnir) [inline test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallProtocolBenchmark.java]: Can we extract this out to a factory class?

**Comment by zoewangg** [comment]
Closing in favor of #1025

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+ * Benchmarking for running with different http clients.
+ */
+@State(Scope.Thread)
+@Warmup(iterations = 3, time = 15, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 5, time = 10, timeUnit = TimeUnit.SECONDS)
+@Fork(2) // To reduce difference between each run
+@BenchmarkMode(Mode.Throughput)
+public class ApiCallHttpClientBenchmark {
+
+    private WireMockServer wireMockServer = new WireMockServer(options().port(PORT_NUMBER));
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java]: Changed to use a local jetty server. From the stack profile, the overhead is negligible.

File: test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    @Param( {"UrlConnectionHttpClient", "ApacheHttpClient"})
+    private String clientType;
+
+    private SdkHttpClient sdkHttpClient;
+
+    private ProtocolRestJsonClient client;
+
+    private Runnable runnable;
+
+    @Setup(Level.Trial)
```
[the remark this answers is not in the document]
Comment (zoewangg) [reply test/sdk-benchmarks/src/main/java/software/amazon/awssdk/benchmark/apicall/ApiCallHttpClientBenchmark.java]: Yeah, I'd still prefer to keep it here though to make it more readable without the need to look at the Javadoc to know what's the default.

**Pull request closed without merging**